main.js

```diff
@@ -0,0 +1,75 @@
+function getRandomInt(min, max) {
+	return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+function randomStep() {
+	const zeroLocation = matrix.findLocation(0);
+	const randomElement = matrix.getRandomStep();
+	if (matrix.exchange(randomElement, zeroLocation))
+		generateTable(matrix);
+}
+
+function locationOffset(location, x, y) {
+	return {
+		x: location.x + x,
+		y: location.y + y
+	}
+}
+
+function onItemClick(e) {
+	const value = parseInt(e.currentTarget.innerText);
+	if (value === 0)
+		return;
+
+	const clickedLocation = matrix.findLocation(value);
+	const emptyLocation = matrix.findLocation(0);
+
+	if (matrix.exchange(clickedLocation, emptyLocation))
+		generateTable(matrix);
+
+	if (matrix.checkWin())
+		setTimeout(function () {
+			alert("Congratulations!!!");
+		}, 0);
```

Comment:
so what do you need this timeout for?

Reply:
it's used to finish rendering (as you mentioned in your previous question)